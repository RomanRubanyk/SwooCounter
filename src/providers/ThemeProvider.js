import React, { useState } from 'react';
import { Appearance } from 'react-native';
import PropTypes from 'prop-types';
import { DARK_COLORS, LIGHT_COLORS } from '../constants/colors';

export const ThemeContext = React.createContext({
  isDark: false,
  colors: LIGHT_COLORS,
  setColorScheme: () => {},
});

export const ThemeProvider = ({ children }) => {
  const colorScheme = Appearance.getColorScheme();
  const [isDark, setIsDark] = useState(colorScheme === 'dark');
  const defaultTheme = {
    isDark,
    colors: isDark ? DARK_COLORS : LIGHT_COLORS,
    setColorScheme: (scheme) => {
      setIsDark(scheme === 'dark');
    },
  };

  return (
    <ThemeContext.Provider value={defaultTheme}>
      {children}
    </ThemeContext.Provider>
  );
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired
};
