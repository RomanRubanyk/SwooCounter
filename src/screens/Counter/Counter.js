import React, { useContext } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/native';
import { Switch } from 'react-native';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import Button from '../../components/Button/Button';
import { setMinusCounterValue, setPlusCounterValue } from '../../store/general/actions';
import { MINUS_STEP, PLUS_STEP } from '../../constants/general';
import { ThemeContext } from '../../providers/ThemeProvider';
import { LIGHT_COLORS, DARK_COLORS } from '../../constants/colors';

const Counter = () => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  const { counterValue } = useSelector(({ general }) => general, shallowEqual);
  const { colors, isDark, setColorScheme } = useContext(ThemeContext);

  const handleChangeColorTheme = (isTrue) => {
    setColorScheme(isTrue ? 'dark' : 'light');
  };

  const handlePlusButton = () => {
    dispatch(setPlusCounterValue(PLUS_STEP));
  };

  const handleMinusButton = () => {
    dispatch(setMinusCounterValue(MINUS_STEP));
  };

  const StyledCounterWrapper = styled.View`
    justify-content: center;
    align-items: center;
    width: 390px;
    height: 844px;
    padding: 222px 115px;
    background-color: ${(props) => (props.isDark ? DARK_COLORS.layout : LIGHT_COLORS.layout)};
`;

  const StyledText = styled.Text`
    width: 160px;
    height: 71px;
    margin-top: 104px;
    margin-bottom: 103px;
    color: #8cd6bd;
    font-weight: 700;
    font-size: 64px;
    letter-spacing: 0;
    text-align: center;
`;

  const Section = styled.View`
    position: absolute;
    align-items: flex-start;
    top: 5px;
    left: 10px;
    padding: 30px 16px;
    background-color: ${(props) => (props.isDark ? DARK_COLORS.layout : LIGHT_COLORS.layout)};
`;

  const SectionText = styled.Text`
    font-size: 16px;
    margin: 12px 0;
    color: ${(props) => (props.isDark ? DARK_COLORS.textColor : LIGHT_COLORS.textColor)};
`;

  return (
    <StyledCounterWrapper isDark={isDark}>
      <Section isDark={isDark}>
        <SectionText isDark={isDark}>
          {isDark ? t('dark') : t('light')}
          {t('mode')}
        </SectionText>
        <Switch
          thumbColor={colors.thumbColor}
          value={isDark}
          onValueChange={handleChangeColorTheme}
        />
      </Section>
      <Button title={`-${MINUS_STEP}`} onPress={handleMinusButton} />
      <StyledText>{counterValue}</StyledText>
      <Button title={`+${PLUS_STEP}`} onPress={handlePlusButton} />
    </StyledCounterWrapper>
  );
};

Counter.propTypes = {
  isDark: PropTypes.bool
};
Counter.defaultProps = {
  isDark: false
};

export default Counter;
