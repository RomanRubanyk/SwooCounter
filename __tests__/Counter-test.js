import React from 'react';
import renderer from 'react-test-renderer';
import Counter from '../src/screens/Counter/Counter';
import { ThemeProvider } from '../src/providers/ThemeProvider';
import store from '../src/store/store';
import { Provider } from 'react-redux';

test('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider>
      <Provider store={store}>
        <Counter />
      </Provider>
    </ThemeProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
