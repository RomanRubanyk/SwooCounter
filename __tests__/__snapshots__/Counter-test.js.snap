// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  isDark={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(0, 71, 97)",
        "height": 844,
        "justifyContent": "center",
        "paddingBottom": 222,
        "paddingLeft": 115,
        "paddingRight": 115,
        "paddingTop": 222,
        "width": 390,
      },
    ]
  }
>
  <View
    isDark={false}
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "rgb(0, 71, 97)",
          "left": 10,
          "paddingBottom": 30,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 30,
          "position": "absolute",
          "top": 5,
        },
      ]
    }
  >
    <Text
      isDark={false}
      style={
        Array [
          Object {
            "color": "rgb(200, 242, 247)",
            "fontSize": 16,
            "marginBottom": 12,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 12,
          },
        ]
      }
    >
      light
      mode
    </Text>
    <RCTSwitch
      accessibilityRole="switch"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 31,
          "width": 51,
        }
      }
      thumbTintColor="rgb(255, 255, 255)"
      value={false}
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 24,
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 160,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "700",
            "letterSpacing": 0,
          },
        ]
      }
    >
      -1
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#8cd6bd",
          "fontSize": 64,
          "fontWeight": "700",
          "height": 71,
          "letterSpacing": 0,
          "marginBottom": 103,
          "marginTop": 104,
          "textAlign": "center",
          "width": 160,
        },
      ]
    }
  >
    0
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderRadius": 24,
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 160,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "700",
            "letterSpacing": 0,
          },
        ]
      }
    >
      +1
    </Text>
  </View>
</View>
`;
